### PR TITLE
Fixes #17073: Skip puppet upgrade if already installed

### PR DESCRIPTION
**Pull request by ehelms**

### CODE REVIEW

**Comment by stbenjam** [comment]
I feel like touching something like `/etc/foreman-installer/.puppet4upgrade` would be safer, a user could upgrade packages and not run the installer and they'll be stuck with p3 param values.

**Comment by ekohl** [comment]
We should also consider puppet 5 migrations. Puppetlabs has published nightly repositories and promised that all code will continue to be runnable under puppet 5. I also think the paths will not change so then a 4 in the filename might be misleading as the current code will also pass for 5. See https://puppet.com/blog/full-visibility-and-control-of-your-infrastructure-new-puppet-releases and skip to Puppet 5 Platform.

**Comment by ehelms** [comment]
Updated per your suggestion @stbenjam

File: hooks/pre/31-upgrade-puppet.rb [hunk 1]
```diff
@@ -5,6 +5,13 @@ def puppet4_available?
   !success.include?(false)
 end
 
+def puppet4_installed?
```
Comment (ekohl) [inline hooks/pre/31-upgrade-puppet.rb]: Is this still used anywhere?

File: hooks/pre/31-upgrade-puppet.rb [hunk 2]
```diff
@@ -74,5 +82,6 @@ def fail_and_exit(message)
   upgrade_step :remove_puppet_port_httpd
   upgrade_step :start_httpd
 
+  File.open('/etc/foreman-installer/.puppet_4_upgrade', 'w') { |file| file.write("completed on #{Time.now}") }
```
Comment (stbenjam) [inline hooks/pre/31-upgrade-puppet.rb]: Based on @ekohl's other comment I think we can change this (and the text below) to just say puppet upgrade to future-proof it so users can go directly from Puppet 3 to 5.

**Comment by ehelms** [comment]
Thanks for the comments, updated

File: hooks/pre/31-upgrade-puppet.rb [hunk 3]
```diff
@@ -47,6 +47,7 @@ def fail_and_exit(message)
 end
 
 if app_value(:upgrade_puppet)
+  fail_and_exit 'Puppet already installed and upgraded. Skipping.' if File.exist?('/etc/foreman-installer/.puppet_4_upgrade')
```
Comment (stbenjam) [inline hooks/pre/31-upgrade-puppet.rb]: This needs updating to the new file name

Comment (ekohl) [inline hooks/pre/31-upgrade-puppet.rb]: I'd prefer if the file location would be a constant somewhere. That removes the risk of mismatches.

**Comment by ehelms** [comment]
@ekohl @stbenjam Updated to make the file a constant

**Pull request merged**